File: 19. Ch-5-2.java
import java.util.*;

class Node 
{
    int key, value;
    Node next;
    Node(int key, int value) 
    {
        this.key = key;
        this.value = value;
        this.next = null;
    }
}

class HashTable 
{
    int SIZE = 10;
    Node[] table;

    HashTable() 
    {
        table = new Node[SIZE];
    }

    int hash(int key) 
    {
        return key % SIZE;
    }

    void insert(int key, int value) 
    {
        int index = hash(key);
        Node newNode = new Node(key, value);

        if (table[index] == null) 
        {
            table[index] = newNode;
        } 
        else 
        {
            Node temp = table[index];
            while (temp.next != null) 
            {
                temp = temp.next;
            }
            temp.next = newNode;
        }
        System.out.println("Inserted: (" + key + "," + value + ")");
    }

    void search(int key) {
        int index = hash(key);
        Node temp = table[index];

        while (temp != null) 
        {
            if (temp.key == key) 
            {
                System.out.println(key + " -> " + temp.value);
                return;
            }
            temp = temp.next;
        }
    }

    void display() {
        for (int i = 0; i < SIZE; i++) 
        {
            System.out.print(i + ": ");
            Node temp = table[i];
            if (temp == null) 
            {
                System.out.println("NULL");
            } 
            else 
            {
                while (temp != null) 
                {
                    System.out.print("(" + temp.key + "," + temp.value + ")-> ");
                    temp = temp.next;
                }
                System.out.println("NULL");
            }
        }
    }
}

public class Main 
{
    public static void main(String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        HashTable ht = new HashTable();

        while (true) 
        {
            int choice = sc.nextInt();
            switch (choice) 
            {
                case 1:
                    int key = sc.nextInt();
                    int value = sc.nextInt();
                    ht.insert(key, value);
                    break;
                case 2:
                    int searchKey = sc.nextInt();
                    ht.search(searchKey);
                    break;
                case 3:
                    ht.display();
                    break;
                case 4:
                    System.out.println("Exit");
                    return;
            }
        }
    }
}

File: 19. Ch-5-2.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None

class HashTable:
    def __init__(self):
        self.size = 10
        self.table = [None] * self.size
    def hash_func(self, key):
        return key % self.size

    def insert(self, key, value):
        index = self.hash_func(key)
        new_node = Node(key, value)
        if self.table[index] is None:
            self.table[index] = new_node
        else:
            curr = self.table[index]
            while curr.next:
                curr = curr.next
            curr.next = new_node
        print(f"Inserted: ({key},{value})")

    def retrieve(self, key):
        index = self.hash_func(key)
        curr = self.table[index]
        while curr:
            if curr.key == key:
                print(f"{key} -> {curr.value}")
                return
            curr = curr.next
    def display(self):
        for i in range(self.size):
            print(f"{i}:", end=" ")
            curr = self.table[i]
            if not curr:
                print("NULL")
            else:
                while curr:
                    print(f"({curr.key},{curr.value})->", end=" ")
                    curr = curr.next
                print("NULL")

ht = HashTable()
while True:
        op = int(input())
        if op == 1:
            key, value = map(int, input().split())
            ht.insert(key, value)
        elif op == 2:
            key = int(input())
            ht.retrieve(key)
        elif op == 3:
            ht.display()
        elif op == 4:
            print("Exit")
            break
